include/nbl/ui/CWindowWin32.h

```diff
@@ -39,97 +41,150 @@ class CWindowWin32 final : public IWindowWin32
 	}
 
 private:
-    CWindowWin32(core::smart_refctd_ptr<system::ISystem>&& sys, uint32_t _w, uint32_t _h, E_CREATE_FLAGS _flags) : IWindowWin32(std::move(sys), _w, _h, _flags), m_native(NULL)
+    CWindowWin32(core::smart_refctd_ptr<system::ISystem>&& sys, uint32_t _w, uint32_t _h, E_CREATE_FLAGS _flags) : 
+		IWindowWin32(std::move(sys), _w, _h, _flags), m_threadHandler(_w, _h, _flags)
     {
-		// get process handle
-		HINSTANCE hinstance = GetModuleHandle(NULL);
-
-		const char* classname = __TEXT("Nabla Engine");
-
-		WNDCLASSEX wcex;
-		wcex.cbSize = sizeof(WNDCLASSEX);
-		wcex.style = CS_HREDRAW | CS_VREDRAW;
-		wcex.lpfnWndProc = WndProc;
-		wcex.cbClsExtra = 0;
-		wcex.cbWndExtra = 0;
-		wcex.hInstance = hinstance;
-		wcex.hIcon = NULL;
-		wcex.hCursor = 0; // LoadCursor(NULL, IDC_ARROW);
-		wcex.hbrBackground = (HBRUSH)(COLOR_WINDOW + 1);
-		wcex.lpszMenuName = 0;
-		wcex.lpszClassName = classname;
-		wcex.hIconSm = 0;
-
-		RegisterClassEx(&wcex);
-
-		// calculate client size
-
-		RECT clientSize;
-		clientSize.top = 0;
-		clientSize.left = 0;
-		clientSize.right = m_width;
-		clientSize.bottom = m_height;
-
-		DWORD style = WS_POPUP; // TODO why popup?
-
-		if (!isFullscreen())
+		m_width = _w; m_height = _h;
+		SetWindowLongPtr(m_threadHandler.getNativeWindow(), GWLP_USERDATA, reinterpret_cast<LONG_PTR>(this));
+    }
+
+
+    //native_handle_t m_native;
+private:
+	class CThreadHandler final : public system::IThreadHandler<CThreadHandler>
+	{
+		using base_t = system::IThreadHandler<CThreadHandler>;
+		friend base_t;
+	public:
+		CThreadHandler(uint32_t _w, uint32_t _h, E_CREATE_FLAGS _flags)
```

Comment:
same as here https://github.com/Crisspl/Nabla/pull/5/commits/ff633e5245b63982c6fe2d57fc56fd659f8fa681#r640991138
You dont want one thread per window, you want one thread managing all win32 windows.

Reply:
there might actually be an upside to having 1 thread : 1 window, being the ability to use `GetMessage` instead of `PeekMessage`.

P.S. Would `WaitForInputIdle` work in a single thread setup?

Reply:
ok so according to your research it seems that `GetMessage` can be configured to wait on all windows, so no downsides to using a single thread.

But if you guys want to use a singleton thread, then you must create the `nbl::ui::IUserInterface` with the associated win32,x11,wayland, and Android implementations for managing singleton/global stuff.

Cause otherwise one would have no clue when to start and kill a global win32 window managment thread.

Also by principle... globals || static members == bad, remember on PC (Win+Lin) we want to enshrine the capability of Nabla to be able to run multiple independent copies of itself within the same process.

Reply:
> Cause otherwise one would have no clue when to start and kill a global win32 window managment thread.

hm good point.. But with IUserInterface being a thing, i still dont know when to destroy the thread. Well, we could say it dies with the last window and then is recreated if we want to create more windows
I dont think its mandatory to create another class for that, but yeah it'll be cleaner. I just wouldnt put this class into user-accessible scope since we dont know how it's gonna be implemented on different platforms (maybe 1 thread per window, maybe no extra threads at all [see Android])

Reply:
Ok, lets put it into user-accesible scope since i dont see other way to not have globals.
Lets name it IWindowManager tho, AFAIK we  have IUI reserved for the future
@sadiuk So whatever is `static` now, put into win32 impl of IWindowManager

Reply:
> Well, we could say it dies with the last window and then is recreated if we want to create more windows

It starts and dies when the `class CUserInterfaceWin32 : public IUserInterface` constructs and destructs.

If you make everything hold a reverse `smart_refctd_ptr<IUserInterface>` that needs the `IUserInterface` to live while it does something, then it will work just like our "life support" of IAPIConnection, IPhysicalDevice, ILogicalDevice, etc. in the video namespace.

> I just wouldnt put this class into user-accessible scope since we dont know how it's gonna be implemented on different platforms (maybe 1 thread per window, maybe no extra threads at all [see Android])

Let `IUserInterface` be a pure virtual interface, with concrete implementations `CUserInterface<Win,Lin,Android>`, then no need for a window manager.